kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ReverseAddRemoveToAddColumnarBatchIterator.java

```diff
@@ -0,0 +1,245 @@
+        }
+
+        // By the definition of `hasNext`, we know that `next` is non-empty
+
+        final DataReadResult ret = next.get();
+        next = Optional.empty();
+        return ret;
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (!closed) {
+            iter.close();
+            closed = true;
+        }
+    }
+
+    /**
+     * Grabs the next FileDataReadResult from `iter` and updates the value of `next`.
+     *
+     * Internally, implements the following algorithm:
+     * 1. read all the RemoveFiles in the next ColumnarBatch to update the `tombstonesFromJson` set
+     * 2. read all the AddFiles in that same ColumnarBatch, unselecting ones that have already
+     *    been removed or returned by updating a selection vector
+     * 3. produces a DataReadResult by dropping that RemoveFile column from the ColumnarBatch and
+     *    using that selection vector
+     *
+     * Ensures that
+     * - `next` is non-empty if there is a next result
+     * - `next` is empty if there is no next result
+     */
+    private void prepareNext() {
+        if (next.isPresent()) return; // already have a next result
+        if (!iter.hasNext()) return; // no next result, and no batches to read
+
+        final Tuple2<FileDataReadResult, Boolean> _next = iter.next();
+        final FileDataReadResult fileDataReadResult = _next._1;
+        final boolean isFromCheckpoint = _next._2;
+        final ColumnarBatch addRemoveColumnarBatch = fileDataReadResult.getData();
+
+        assert(addRemoveColumnarBatch.getSchema().equals(LogReplay.ADD_REMOVE_READ_SCHEMA));
+
+        // Step 1: Update `tombstonesFromJson` with all the RemoveFiles in this columnar batch, if
+        //         and only if this batch is not from a checkpoint.
+        //
+        //         There's no reason to put a RemoveFile from a checkpoint into `tombstonesFromJson`
+        //         since, when we generate a checkpoint, any corresponding AddFile would have
+        //         been excluded already
+        if (!isFromCheckpoint) {
+            final ColumnVector removesVector = addRemoveColumnarBatch.getColumnVector(1);
+            for (int rowId = 0; rowId < removesVector.getSize(); rowId++) {
+                if (removesVector.isNullAt(rowId)) continue;
+
+                // Note: this row doesn't represent the complete RemoveFile schema. It only contains
+                //       the fields we need for this replay.
+
+                final Row row = removesVector.getStruct(rowId);
+                final String path = requireNonNull(row, 0, "path").getString(0);
+                final URI pathAsUri = pathToUri(path);
+                final Optional<String> dvId = Optional.ofNullable(
+                    DeletionVectorDescriptor.fromRow(row.getStruct(1))
+                ).map(DeletionVectorDescriptor::getUniqueId);
+                final UniqueFileActionTuple key = new UniqueFileActionTuple(pathAsUri, dvId);
```

Comment:
i think we might have a problem here. since the file actions are not ordered in a json file... there might be problems. 

5.json
remove(f1, d1)
add(f1, d2)

is same as 

5.json
remove(f1, d1)
add(f1, d2)


are you sure your logic is handling this?

Reply:
In step 1, we load all the RemoveFiles into the hash set.

In step 2, we use that hash set to update the selection vector for the add files.

I don't see any dependency here on the order of adds and removes within the JSON file. I am confident my logic is handling this.

Reply:
hm. but this is assuming each JSON file is not split.

```
00005.json___split01
AddA --> this would be selected, since RemoveA is not yet seen.

00005.json___split01
RemoveA
```

Reply:
Is it acceptable to require that JSON files are not split?

Reply:
Nvm. not an issue. 005.json will never have RemoveA and AddA together, regardless of splits.